First commit of a mini-book on building programs (#90)
fortran-lang/fortran-lang.org: learn/building_programs.md

---
layout: page
title: Building programs
permalink: /learn/build_programs
navbar: Learn
---

> Draft tutorial: Building programs
> ---------------------------------

Reviewer: ```suggestion
```
This won't be needed, the `title` field in the header will be rendered first as a H1 heading.

> Remark: I only included Linux and Windows as operating systems in this
> tutorial, for the simple reason that I have no experience with MacOS*
> in any of its guises. While MacOS* is very similar to Linux, it is the
> differences that may need to be documented. Some information would be
> helpful.

Reviewer: ```suggestion
Remark: this tutorial gives examples for the Windows and Linux operating systems,
however the workflow and general principles still apply to MacOS.
```

It would be good if someone with experience developing on MacOS is able to check the tutorial content for any significant differences, especially with libraries.
If not, then maybe we can link to a good online resource for this?

Author: Arjen Markus

dd. 22 may 2020

Introduction: compiled languages
================================

Languages like Fortran, C, C++ and Java, to name but a few, share
certain characteristics: you write code in your language of choice but
then you have to build an executable program from that source code.
Other languages are interpreted - the source code is analysed by a
special program and taken as direct instructions. Two very simple
examples of that type of language: Windows batch files and Linux shell
scripts.

In this tutorial we concentrate on the first type of languages, with
Fortran as the main example. One advantage of compiled languages is that
the build process that you need to build an executable program, is used
to transform the human-readable source code into an efficient program
that can be run on the computer.

Let us have a look at a simple example:

>     program hello
>         write(*,*) 'Hello!'
>     end program hello

Reviewer 2: ```suggestion
```fortran
    program hello
        write(*,*) 'Hello!'
    end program hello
```
```

where they are found.

If you want to share your program with colleagues or clients or simply
users all over the world, you will have to take care that, besides the
program, you also distribute the libraries it depends on. For more
information: see below.

Include files and modules
-------------------------

Your program can be contained in various source files, all stored in the
same directory or organised in some convenient directory tree. The
details of the organisation depend on personal taste, arrangements made
by the group of developers you belong to, or simply the history of the
program. Whatever the directory structure, you will encounter a
situation where the compiler needs assistence in order to compile a
particular source file:

* Fortran (and other languages) has the possibility to include an
external file. While this feature has become a bit less useful with the
advent of modules, it still has its uses. Quite often, such "include
files" are stored in a directory separated from the directories
containing the source files, because they are used in several locations.
Quite often the name of that directory is "include".

> * As we have seen, source code that defines one or more modules, leads to
> so-called "module intermediate files" (with the extension ".mod"). The

Reviewer: ```suggestion
* As we have seen, compiling source code that defines one or more modules, leads to
the compiler generating so-called "module intermediate files" (with the extension ".mod"). The
```

files and module intermediate files are to be found. Suppose we store
the two files of our tabulation program in the following directory
structure:

    tabulate/
        main/
            tabulate.f90
        sub/
            function.f90

Compiling the file "function.f90" with the commands

    $ cd sub
    $ gfortran -c function.f90

leads to this structure:

        tabulate/
            main/
                tabulate.f90
            sub/
                function.f90
                function.mod
                function.o

To successfully compile and subsequently build the program we need to
tell the compiler where it can find the file "function.mod":

    $ cd main
    $ gfortran -c tabulate.f90 -I ../sub
    $ gfortran -o tabulate tabulate.o ../sub/function.o

The result:

    tabulate/
        main/
            tabulate.f90
            tabulate.o
            tabulate (or tabulate.exe on Windows)
        sub/
            function.f90
            function.mod
            function.o

Notes:

* The details differ per compiler. Sometimes the "-I" option should be
followed by a space and then the name of the directory, sometimes the
directory should come consecutively.

> * Many compilers allow you to specify the location for the module

Reviewer: This is a little too ambiguous for a beginner; can you make the difference between the `-I` flag and the `-J` flag clearer.
_i.e._ that compilers place the .mod files in the current directory by default and that the `-J` flag explicitly specifies an alternative location to place them (and search for them).

    $ ar r supportlib.a file1.o file2.o
    $ ar r supportlib.a file3.o ...

or, using the "link" utility:

    c:\...> ifort -c file1.f90 file2.f90
    c:\...> ifort -c file3.f90 ...
    c:\...> lib /out:supportlib.lib file1.obj file2.obj
    c:\...> link supportlib.lib file3.obj ...

Note:

* The command "ar" with the option "r" either creates the library (the
name appears after the option) or adds new object files to the library
(or replaces any existing ones).
* The command "lib" will create a new library if you use specify the
option "/out:" with the name of the new library next to it. To add
object files to an existing library, leave out the "/out:" bit.
* On platforms like Linux there is a particular convention to name
libraries. If you name your library like "libname.a" (note the "lib"
prefix), then you can refer to it as "-lname" in the link step.
* Libraries are often sought in directories indicated by an option "-L"
or "/LIBPATH". This saves you from having to specify the exact path for
every library.

Using libraries you can build very large programs without
having to resort to extremely long command lines.

### Static versus dynamic libraries

> The above discussion is tacitly assuming that you are using so-called

Reviewer 2: ```suggestion
The above discussion is tacitly assuming that you are using the so-called

```

static libraries. Static libraries (or at least parts of their
contents) become an integral part of the executable program. The only
way to change the routines incorporated in the program is by rebuilding
the program with a new version of the library.

> A flexible alternative is to use so-called dynamic libraries. These

Reviewer 2: ```suggestion
A flexible alternative is to use the so-called dynamic libraries. These

```

> libraries remain outside the executable program and as a consequence you

Reviewer 2: ```suggestion
libraries remain outside the executable program and as a consequence

```

> can replace them without rebulding the entire program. Compilers and

Reviewer 2: ```suggestion
can be replaced without rebulding the entire program. Compilers and

```

indeed the operating system itself rely heavily on such dynamic
libraries. You could consider dynamic libraries as a sort of executable
programs that need a bit of help to be run.

> Building dynamic libraries works slightly differently:

Reviewer 2: Differently based on what? I have the feeling that something is missing in this sentence.

On Linux:

    $ gfortran -fpic -c file1.f90 file2.f90
    $ gfortran -fpic -c file3.f90 ...
    $ gfortran -shared --o supportlib.so file1.o file2.o file3.o ...

On Windows, with the Intel Fortran compiler:

    $ ifort -c file1.f90 file2.f90
    $ ifort -c file3.f90 ...
    $ ifort -dll -exe:supportlib.dll file1.obj file2.obj file3.obj ...

> The differences:

Reviewer 2: ```suggestion
The differences are that:

```

> * You need to specify a compile option on Linux, "-fpic", because the

Reviewer 2: This option is for gfortran right?It could be good to mention it explicitely

object code is slightly different.
* You need to tell in the link step that you want a dynamic library (on
Linux: a shared object/library, hence the extension ".so"; on Windows:
a dynamic link library)

There is one more thing to be aware of: On Windows you must

> explicitly specify that a routine is to be _exported_, i.e. is visible

Reviewer 2: ```suggestion
explicitly specify that a procedure is to be _exported_, i.e. is visible

```

in the dynamic library. There are several ways - depending on the
compiler you use - to achieve this. One method is via a so-called
compiler directive:

    subroutine myroutine( ... )
    !GCC$ ATTRIBUTES DLLEXPORT:: myroutine

Or, with the Intel Fortran compiler:

    subroutine myroutine( ... )
    !DEC$ ATTRIBUTES DLLEXPORT:: myroutine

Besides a dynamic library (DLL), a so-called import library may be
generated.

Because the details differ per compiler, here are two examples:
gfortran on Cygwin and Intel Fortran on Windows. In both cases

> we look at the tabulation program.

Reviewer 2: ```suggestion
we look at the `tabulation.f90` program.

```

> ### Cygwin and gfortran

Reviewer 2: It could be good to replace Cygwin by GNU/Linux

> The "tabulate" program requires a user-defined routine "f". If we

Reviewer 2: ```suggestion
The `tabulate` program requires a user-defined routine `f`. If we

```

let it reside in a dynamic library, say "function.dll", we can simply
replace the implementation of the function by putting another dynamic
library in the directory. No need to rebuild the program as such.

> On Cygwin it is not necessary to explicitly export a routine - all

Reviewer 2: ```suggestion
On Cygwin it is not necessary to explicitly export a procedure - all

```

> publically visible routines are exported when you build a dynamic library.

Reviewer 2: ```suggestion
publically visible procedures are exported when you build a dynamic library.

```

Also, no import library is generated.

Since our dynamic library can be built from a single source file, we
can take a shortcut:

    $ gfortran -shared -o function.dll function.f90

This produces the files "function.dll" and "function.mod". The

> utility "nm" tells us the exact name of the function "f":

Reviewer 2: ```suggestion
utility "nm" tells us the exact name of the function `f`:

```

    $ nm function.dll
    ...
    000000054f9d7000 B __dynamically_loaded
                     U __end__
    0000000000000200 A __file_alignment__
    000000054f9d1030 T __function_MOD_f
    000000054f9d1020 T __gcc_deregister_frame
    000000054f9d1000 T __gcc_register_frame
    ...

It has received a prefix `__function_MOD_` to distinguish it from any

> other routine "f" that might be defined in another module.

Reviewer 2: ```suggestion
other procedure `f` that might be defined in another module.

```

name for the executable - here we do without the "-exe" option.)

Just as under Cygwin, the DLL and the .mod file are used to build the executable program
with checks on the function's interface, the right name and the reference
to "a" DLL, called "function.dll".

You can replace the shared library "function.dll" by another one, but the same
caution is required: while the implementation can be quite different, the
function's interface must be the same.

Build tools
-----------

If this seems complicated, well, you are right and we are only
scratching the surface here. The complications arise because of
differences between platforms, differences between compilers/linkers and
because of differences in the way programs are set up. Fortunately,
there are many tools to help configure and maintain the build steps.
We will not try and catalogue them, but give instead a very limited
list of tools that you typically encounter:

* The "make" utility is a classical tool that uses instructions about
how the various components of a program depend on each other to
efficiently compile and link the program (or programs). It takes a

> so-called makefile that contains the dependencies.

Reviewer 2: ```suggestion
so-called `Makefile` that contains the dependencies.

```

    Simply put:

    If a program file is older than any of the libraries and object files
it depends on, the make utility knows it has to rebuild it and goes on
to look at the libraries and object files - are any out of date?

    If an object file is older than the corresponding source file, the
make utility knows it has to compile the source file.

* Integrated development tools take care of many of the above details. A
popular tool on Windows is MicroSoft's Visual Studio, but others exist,

> such Eclipse (Photran) and Code::Blocks. They offer a graphical

Reviewer 2: ```suggestion
such as Eclipse (Photran) and Code::Blocks. They offer a graphical

```

user-interface, but are often very specific for the compiler and
platform.

* Maintenance tools like autotools and CMake can generate makefiles or
Visual Studio project files via a high-level description. They abstract
away from the compiler and platform specifics.

> Here is a very simple example of a makefile as used by the make utility,

Reviewer 2: ```suggestion
Here is a very simple example of a `Makefile` as used by the make utility,

```

3. Distribute static or dynamic libraries that people can use

ad 1.

By far the simplest - for you as a programmer - is this one: you leave it
up to the user to build it on their own machine. Unfortunately, that
means you will have to have a user-friendly build system in place and
the user will have to have access to suitable compilers. For build systems:
see the previous section.

ad 2.

A pre-built program that does not need to be customised, other than via its
input, will still need to come with the various run-time libraries and will
be specific to the operating system/environment it was built for.

The set of run-time libraries differs per operating system and compiler version.
For a freely available compiler like gfortran, the easiest thing is to ask the
user to install that compiler on their system. In the case of Windows: the Cygwin
environment may be called for.

Alternatively, you can supply copies of the run-time libraries together with your
program. Put them in the directory where they can be found at run-time.

Note: On Windows, the Intel Fortran comes with a set of _redistributable_ libraries.
These will need to be made available.

> In general: use a tool like "ldd" or "dependency walker" to find out what

Reviewer 2: ```suggestion
In general: use a tool like `ldd` or `dependency walker` to find out what

```
